---
layout: default
title: "Methylaction: How-To"
---

<script src="methylaction_files/jquery-1.11.0/jquery.min.js"></script>
<!-- <meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="methylaction_files/bootstrap-3.3.1/css/bootstrap.min.css" rel="stylesheet" />
<script src="methylaction_files/bootstrap-3.3.1/js/bootstrap.min.js"></script>
<script src="methylaction_files/bootstrap-3.3.1/shim/html5shiv.min.js"></script>
<script src="methylaction_files/bootstrap-3.3.1/shim/respond.min.js"></script> -->


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #f8f8f8; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
pre, code { background-color: #f8f8f8; }
code > span.kw { color: #204a87; font-weight: bold; }
code > span.dt { color: #204a87; }
code > span.dv { color: #0000cf; }
code > span.bn { color: #0000cf; }
code > span.fl { color: #0000cf; }
code > span.ch { color: #4e9a06; }
code > span.st { color: #4e9a06; }
code > span.co { color: #8f5902; font-style: italic; }
code > span.ot { color: #8f5902; }
code > span.al { color: #ef2929; }
code > span.fu { color: #000000; }
code > span.er { font-weight: bold; }
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }


  .table td {
   text-align: center;   
}

.table th {
   text-align: center;   
}
</style>


<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img { 
  max-width:100%; 
  height: auto; 
}
</style>
<div class="container-fluid main-container">


<div id="header">
<h1 class="title">Methylaction: Detecting differentially methylated regions (DMRs) that distinguish disease subtypes</h1>
<h4 class="author"><em>Jeff Bhasin</em></h4>
<h4 class="date"><em>2015-05-07</em></h4>
</div>

<div id="TOC">
<ul>
<li><a href="#purpose">Purpose</a></li>
<li><a href="#prerequisites">Prerequisites</a></li>
<li><a href="#installation">Installation</a></li>
<li><a href="#obtaining-the-example-data">Obtaining the Example Data</a></li>
<li><a href="#preprocessing">Preprocessing</a><ul>
<li><a href="#describing-the-experimental-design">Describing the Experimental Design</a></li>
<li><a href="#loading-reads-and-generating-count-tables">Loading Reads and Generating Count Tables</a></li>
</ul></li>
<li><a href="#differentially-methylated-region-dmr-detection">Differentially Methylated Region (DMR) Detection</a><ul>
<li><a href="#interpreting-the-results">Interpreting the results</a></li>
</ul></li>
<li><a href="#vizualization-of-dmrs">Vizualization of DMRs</a></li>
<li><a href="#permutation-and-bootstrap-testing">Permutation and Bootstrap Testing</a></li>
</ul>
</div>

<div id="purpose" class="section level1">
<h1>Purpose</h1>
<p>This how-to will demonstrate the use of Methylaction to detect differentially methylated regions (DMRs) among three groups using data from MBD-isolated Genome Sequencing (MiGS). While Methylaction is designed for genome-wide analysis, this example data is only for a subset of the genome and a subset of samples so the example can be worked through quickly. Please refer to the function documentation for advanced options.</p>
</div>
<div id="prerequisites" class="section level1">
<h1>Prerequisites</h1>
<ul>
<li>R Installed (download from <a href="http://cran.r-project.org/" class="uri">http://cran.r-project.org/</a>)</li>
<li>A linux/unix-based server or workstation (includes Mac OS X)</li>
</ul>
<p>RAM and CPU requirements will depend on the depth of the sequencing and the number of samples and groups. We recommend very high performance machines. As a reference, we used a linux server with 20 cores of 2.80GHz CPUs and 64GB of RAM for a genome-wide analysis of 22 samples across 3 groups. A high performance computing cluster (HPC) was used to obtain 1,000s of permutations.</p>
<p>For the purposes of the example, less CPU and RAM are required. This how-to has been tested on a MacBook Pro with a 2.4 GHz CPU and 8 GB of RAM.</p>
</div>
<div id="installation" class="section level1">
<h1>Installation</h1>
<p>First, start R (by typing the “R” command at the linux command line) and install pre-requisite packages from <a href="http://www.bioconductor.org">Bioconductor</a>:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">source</span>(<span class="st">&quot;http://bioconductor.org/biocLite.R&quot;</span>)
<span class="kw">biocLite</span>(<span class="kw">c</span>(<span class="st">&quot;GenomicRanges&quot;</span>,<span class="st">&quot;IRanges&quot;</span>,<span class="st">&quot;devtools&quot;</span>,<span class="st">&quot;DESeq&quot;</span>,<span class="st">&quot;GenomicAlignments&quot;</span>,
           <span class="st">&quot;Repitools&quot;</span>,<span class="st">&quot;Rsubread&quot;</span>,<span class="st">&quot;ggbio&quot;</span>))</code></pre>
<p>Then, install both Goldmine and Methylaction from GitHub. Be sure to accept installation of any additional pre-requisite packages from CRAN.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(devtools)
<span class="kw">install_github</span>(<span class="st">&quot;jeffbhasin/goldmine&quot;</span>)
<span class="kw">install_github</span>(<span class="st">&quot;jeffbhasin/methylaction&quot;</span>)</code></pre>
</div>
<div id="obtaining-the-example-data" class="section level1">
<h1>Obtaining the Example Data</h1>
<p>Please obtain the “methylaction_demo.tar.gz” file and extract it. This contains all the example input data needed to complete this how-to.</p>
<p>From the linux command line, use the command “tar -zxvf methylaction_demo.tar.gz” for extraction.</p>
<p>Note that your session of R must have the working directory set to the location of the “methylaction_demo” folder extracted from the tar.gz file so paths to files in this how-to will be valid. Use the “setwd()” function to set this directory.</p>
<p>For example (change to your location):</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">setwd</span>(<span class="st">&quot;/Users/bhasinj/Documents/methylaction_demo&quot;</span>)</code></pre>
</div>
<div id="preprocessing" class="section level1">
<h1>Preprocessing</h1>
<p>First, start R, and load the methylaction R package into the session.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(methylaction)</code></pre>
<div id="describing-the-experimental-design" class="section level2">
<h2>Describing the Experimental Design</h2>
<p>Unique sample identifiers, paths to BAM files, group assignments, and any sample covariates are defined in a CSV file. Please see “input/samples.csv” for an example of the format. Note that comparison groups must be ordered as you want the groups to appear in the output results. This is important, because in the output, patterns between the groups will be coded using sequences of binary digits, where each digit represents a group, and this is the same as the order the groups are encountered in the sample CSV file. Optionally, a column called “color” can be provided that defines colors for each groups. These will be used in certain plotting and reporting functions. If this column is omitted, colors are automatically assigned using RColorBrewer.</p>
<p>The sample CSV file can be read into R using the readSampleInfo() command:</p>
<pre class="sourceCode r"><code class="sourceCode r">samp &lt;-<span class="st"> </span><span class="kw">readSampleInfo</span>(<span class="st">&quot;input/samples.csv&quot;</span>)</code></pre>
<pre><code>## Found 9 samples in 3 groups
## Group order for pattern and contrast digits will be: benign, low, high
## Number of samples in each group:</code></pre>
<pre><code>## benign    low   high 
##      3      3      3</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(samp)</code></pre>
<pre><code>##   sample  group                 bam ancestry   color
## 1    BE1 benign input/BE1.chr22.bam european #4daf4a
## 2    BE2 benign input/BE2.chr22.bam european #4daf4a
## 3    BE3 benign input/BE3.chr22.bam european #4daf4a
## 4    LE1    low input/LE1.chr22.bam european #377eb8
## 5    LE2    low input/LE2.chr22.bam european #377eb8
## 6    LE3    low input/LE3.chr22.bam european #377eb8
## 7    HE1   high input/HE1.chr22.bam european #e41a1c
## 8    HE2   high input/HE2.chr22.bam european #e41a1c
## 9    HE3   high input/HE3.chr22.bam european #e41a1c</code></pre>
<p>The command will output confirmation of group sample sizes and group order.</p>
</div>
<div id="loading-reads-and-generating-count-tables" class="section level2">
<h2>Loading Reads and Generating Count Tables</h2>
<p>All read alignments are read and processed from the BAM files into an RData workspace that is saved to disk, and this prevents the user from having to re-process data again for each run of Methylaction. The getReads() function obtains the coordinates of the read alignments from the BAM file. The initial stage of the program works with fragment counts in 50bp windows, which are also computed once and saved by the function getCounts().</p>
<p>There are two other variables specific to the experiment that are needed in the next step. These are the window size (we recommend 50bp) and the fragment size used for generating the sequencing library. This fragment size is not the read length – it is the average size of a fragment in the sequencing library that was prepared. Often this number is available from BioAnayzer results, and should be known to whomever did the sequencing library preparation for your study. In the case of paired end data, the fragsize variable can be set to “paired” and valid mate pairs will be used.</p>
<p>First, define the variables that will be needed for the preprocessing functions:</p>
<pre class="sourceCode r"><code class="sourceCode r">chrs &lt;-<span class="st"> &quot;chr22&quot;</span>
fragsize &lt;-<span class="st"> </span><span class="dv">120</span>
winsize &lt;-<span class="st"> </span><span class="dv">50</span>
ncore &lt;-<span class="st"> </span><span class="dv">1</span></code></pre>
<p>Then, read in alignments and generate binned count tables (these steps may be memory, disk, and CPU intensive):</p>
<pre class="sourceCode r"><code class="sourceCode r">reads &lt;-<span class="st"> </span><span class="kw">getReads</span>(<span class="dt">samp=</span>samp, <span class="dt">chrs=</span>chrs, <span class="dt">fragsize=</span>fragsize, <span class="dt">ncore=</span>ncore)
counts &lt;-<span class="st"> </span><span class="kw">getCounts</span>(<span class="dt">samp=</span>samp, <span class="dt">reads=</span>reads, <span class="dt">chrs=</span>chrs, <span class="dt">winsize=</span>winsize, <span class="dt">ncore=</span>ncore)</code></pre>
<p>We recommend saving all of the above into a single preprocessing RData, which can be loaded prior to running the DMR detection step described next.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">save</span>(samp, reads, counts, winsize, fragsize, chrs, <span class="dt">file=</span><span class="st">&quot;output/prepro.rd&quot;</span>,<span class="dt">compress=</span>T)</code></pre>
<p>For future runs of DMR detection, the saved RData can be loaded rather than spending time re-preprocessing the data.</p>
</div>
</div>
<div id="differentially-methylated-region-dmr-detection" class="section level1">
<h1>Differentially Methylated Region (DMR) Detection</h1>
<p>With all the preprocessing completed, differentially methylated regions (DMRs) can be detected using a call to the methylaction() function. There are many options to this function that will affect the DMR detection. Here we have used recommended defaults. See the function documentation for more details. This function performs multiple steps, which are all CPU, RAM, and disk intensive when run on larger data sets.</p>
<p>First, set the number of cores to use based on your hardware (we recommend reducing this number if there are memory issues):</p>
<pre class="sourceCode r"><code class="sourceCode r">ncore &lt;-<span class="st"> </span><span class="dv">1</span></code></pre>
<p>Then, run methylaction() to call DMRs:</p>
<pre class="sourceCode r"><code class="sourceCode r">ma &lt;-<span class="st"> </span><span class="kw">methylaction</span>(<span class="dt">samp=</span>samp, <span class="dt">counts=</span>counts, <span class="dt">reads=</span>reads, <span class="dt">ncore=</span>ncore)</code></pre>
<p>Finally, save the results object to disk:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">save</span>(ma,<span class="dt">file=</span><span class="st">&quot;output/ma.rd&quot;</span>, <span class="dt">compress=</span>T)</code></pre>
<p>It is also useful to save a DMR list as a CSV file for viewing in a spreadsheet application.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">write.csv</span>(<span class="kw">makeDT</span>(ma$dmr), <span class="dt">row.names=</span><span class="ot">FALSE</span>, <span class="dt">file=</span><span class="st">&quot;dmrs.csv&quot;</span>)</code></pre>
<div id="interpreting-the-results" class="section level2">
<h2>Interpreting the results</h2>
<p>The function methylaction() is designed to output a great deal of information about the internals of the DMR calling in order to facilitate comparisons between different settings and to prevent needing to re-run the command on large datasets to view intermediate states. The output object is a list. For a list of all detected DMRs, look at ma$dmr. Each DMR is assigned a pattern code, where each digit represents a group. The pattern indicates the differential methylation status between the groups. All groups with the digit is 1 are hypermethylated with respect to all groups with the digit 0. See the ma$args object for the arguments provided to the call of methylaction() that generated the output. If you want to access data from any internal steps of the function, see the objects nested under ma$data.</p>
<p>The maSummary() function provides details about the intermediate stages of the method.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">maSummary</span>(ma)</code></pre>
<pre><code>##                                           stat   count percent
## 1                                  Window Size      50        
## 2                                Total Windows 1026091        
## 3                  All Zero Windows (filtered)  440064   42.89
## 4             All Below FDR Windows (filtered)  409356   39.89
## 5        Signal Windows (move on to stage one)  176671   17.22
## 6                  Windows Tested in Stage One  176671        
## 7                     Sig Pattern in Stage One  158274   89.59
## 8                 Non-Sig Pattern in Stage One       0       0
## 9                   Ambig Pattern in Stage One   18397   10.41
## 10 Regions Formed By Joining Adjacent Patterns    1351        
## 11                 Regions Tested in Stage Two    1351        
## 12                    Regions That Pass ANODEV    1330   98.45
## 13                 ANODEV Sig with Sig Pattern    1263   94.96
## 14             ANODEV Sig with Non-sig Pattern       0       0
## 15               ANODEV Sig with Ambig Pattern      67    5.04
## 16                                  Total DMRs    1263</code></pre>
<p>The built-in function table() can be used to obtain counts of DMRs by pattern and “frequent” classification.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(ma$dmr$pattern,ma$dmr$frequent)</code></pre>
<table>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">FALSE</th>
<th align="right">TRUE</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">001</td>
<td align="right">41</td>
<td align="right">53</td>
</tr>
<tr class="even">
<td align="left">010</td>
<td align="right">465</td>
<td align="right">21</td>
</tr>
<tr class="odd">
<td align="left">011</td>
<td align="right">98</td>
<td align="right">86</td>
</tr>
<tr class="even">
<td align="left">100</td>
<td align="right">80</td>
<td align="right">12</td>
</tr>
<tr class="odd">
<td align="left">101</td>
<td align="right">241</td>
<td align="right">72</td>
</tr>
<tr class="even">
<td align="left">110</td>
<td align="right">87</td>
<td align="right">7</td>
</tr>
<tr class="odd">
<td align="left">ambig</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="vizualization-of-dmrs" class="section level1">
<h1>Vizualization of DMRs</h1>
<p>DMRs can be visualized genome-wide via a heatmap or karyogram.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">maHeatmap</span>(ma)</code></pre>
<p><img src="methylaction_files/figure-html/unnamed-chunk-16-1.png" title="" alt="" width="672" /></p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">maKaryogram</span>(<span class="dt">ma=</span>ma, <span class="dt">reads=</span>reads)</code></pre>
<p><img src="methylaction_files/figure-html/unnamed-chunk-17-1.png" title="" alt="" width="576" /></p>
<p>In this case, the Karyogram only spans chr22 due to the subset of data for this example.</p>
<p>It is also useful to view DMRs and the windowed sequencing read counts in a genome browser, such as the UCSC genome browser. The maBed() and maTracks() functions can create BED files suitable for uploading as custom tracks to <a href="http://genome.ucsc.edu/" class="uri">http://genome.ucsc.edu/</a>.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">maBed</span>(ma,<span class="dt">file=</span><span class="st">&quot;dmrs.bed&quot;</span>)
<span class="kw">maTracks</span>(<span class="dt">ma=</span>ma, <span class="dt">reads=</span>reads, <span class="dt">path=</span><span class="st">&quot;beds&quot;</span>)</code></pre>
</div>
<div id="permutation-and-bootstrap-testing" class="section level1">
<h1>Permutation and Bootstrap Testing</h1>
<p>Because of the two stage testing approach, type I error rates may be inflated with this method. To determine if this level is acceptable, we have implemented permutation and bootstrapping approaches. This establishes a false discovery rate (FDR) for each pattern of DMR among the groups. If these FDRs are too high, they can be recalculated at lower p-value thresholds until they reach acceptable levels. Then, DMRs filtered at this p-value can be used as the definitive list for the study.</p>
<p>Bootstraps can be enabled by adding the “nperms” and “perm.boot” options to methylaction(). The resulting output list will then have an “fdr” object that reports false discovery rates (FDRs).</p>
<pre class="sourceCode r"><code class="sourceCode r">ma &lt;-<span class="st"> </span><span class="kw">methylaction</span>(<span class="dt">samp=</span>samp, <span class="dt">counts=</span>counts, <span class="dt">reads=</span>reads, <span class="dt">perm.boot=</span>T, <span class="dt">nperms=</span><span class="dv">3</span>, <span class="dt">ncore=</span>ncore)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(ma$fdr)</code></pre>
<table>
<thead>
<tr class="header">
<th align="right">pattern</th>
<th align="right">type</th>
<th align="right">nDMRs</th>
<th align="right">permMean</th>
<th align="right">permSD</th>
<th align="right">permCV</th>
<th align="right">FDRpercent</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">001</td>
<td align="center">all</td>
<td align="center">94</td>
<td align="center">31.33</td>
<td align="center">18.56</td>
<td align="center">0.59</td>
<td align="center">33.33</td>
</tr>
<tr class="even">
<td align="left">001</td>
<td align="left">frequent</td>
<td align="right">53</td>
<td align="right">3.33</td>
<td align="right">3.51</td>
<td align="right">1.05</td>
<td align="right">6.29</td>
</tr>
<tr class="odd">
<td align="left">001</td>
<td align="left">other</td>
<td align="right">41</td>
<td align="right">28.00</td>
<td align="right">17.06</td>
<td align="right">0.61</td>
<td align="right">68.29</td>
</tr>
<tr class="even">
<td align="left">010</td>
<td align="left">all</td>
<td align="right">486</td>
<td align="right">6.33</td>
<td align="right">6.03</td>
<td align="right">0.95</td>
<td align="right">1.30</td>
</tr>
<tr class="odd">
<td align="left">010</td>
<td align="left">frequent</td>
<td align="right">21</td>
<td align="right">2.67</td>
<td align="right">3.79</td>
<td align="right">1.42</td>
<td align="right">12.70</td>
</tr>
<tr class="even">
<td align="left">010</td>
<td align="left">other</td>
<td align="right">465</td>
<td align="right">3.67</td>
<td align="right">3.21</td>
<td align="right">0.88</td>
<td align="right">0.79</td>
</tr>
<tr class="odd">
<td align="left">011</td>
<td align="left">all</td>
<td align="right">184</td>
<td align="right">255.33</td>
<td align="right">379.16</td>
<td align="right">1.48</td>
<td align="right">138.77</td>
</tr>
<tr class="even">
<td align="left">011</td>
<td align="left">frequent</td>
<td align="right">86</td>
<td align="right">79.00</td>
<td align="right">128.27</td>
<td align="right">1.62</td>
<td align="right">91.86</td>
</tr>
<tr class="odd">
<td align="left">011</td>
<td align="left">other</td>
<td align="right">98</td>
<td align="right">176.33</td>
<td align="right">251.24</td>
<td align="right">1.42</td>
<td align="right">179.93</td>
</tr>
<tr class="even">
<td align="left">100</td>
<td align="left">all</td>
<td align="right">92</td>
<td align="right">287.00</td>
<td align="right">495.37</td>
<td align="right">1.73</td>
<td align="right">311.96</td>
</tr>
<tr class="odd">
<td align="left">100</td>
<td align="left">frequent</td>
<td align="right">12</td>
<td align="right">91.67</td>
<td align="right">158.77</td>
<td align="right">1.73</td>
<td align="right">763.89</td>
</tr>
<tr class="even">
<td align="left">100</td>
<td align="left">other</td>
<td align="right">80</td>
<td align="right">195.33</td>
<td align="right">336.60</td>
<td align="right">1.72</td>
<td align="right">244.17</td>
</tr>
<tr class="odd">
<td align="left">101</td>
<td align="left">all</td>
<td align="right">313</td>
<td align="right">22.67</td>
<td align="right">34.20</td>
<td align="right">1.51</td>
<td align="right">7.24</td>
</tr>
<tr class="even">
<td align="left">101</td>
<td align="left">frequent</td>
<td align="right">72</td>
<td align="right">2.67</td>
<td align="right">2.52</td>
<td align="right">0.94</td>
<td align="right">3.70</td>
</tr>
<tr class="odd">
<td align="left">101</td>
<td align="left">other</td>
<td align="right">241</td>
<td align="right">20.00</td>
<td align="right">32.08</td>
<td align="right">1.60</td>
<td align="right">8.30</td>
</tr>
<tr class="even">
<td align="left">110</td>
<td align="left">all</td>
<td align="right">94</td>
<td align="right">216.67</td>
<td align="right">206.29</td>
<td align="right">0.95</td>
<td align="right">230.50</td>
</tr>
<tr class="odd">
<td align="left">110</td>
<td align="left">frequent</td>
<td align="right">7</td>
<td align="right">61.00</td>
<td align="right">38.97</td>
<td align="right">0.64</td>
<td align="right">871.43</td>
</tr>
<tr class="even">
<td align="left">110</td>
<td align="left">other</td>
<td align="right">87</td>
<td align="right">155.67</td>
<td align="right">167.46</td>
<td align="right">1.08</td>
<td align="right">178.93</td>
</tr>
<tr class="odd">
<td align="left">all</td>
<td align="left">all</td>
<td align="right">1263</td>
<td align="right">819.33</td>
<td align="right">1063.48</td>
<td align="right">1.30</td>
<td align="right">64.87</td>
</tr>
<tr class="even">
<td align="left">all</td>
<td align="left">frequent</td>
<td align="right">251</td>
<td align="right">240.33</td>
<td align="right">320.38</td>
<td align="right">1.33</td>
<td align="right">95.75</td>
</tr>
<tr class="odd">
<td align="left">all</td>
<td align="left">other</td>
<td align="right">1012</td>
<td align="right">579.00</td>
<td align="right">743.21</td>
<td align="right">1.28</td>
<td align="right">57.21</td>
</tr>
</tbody>
</table>
<p>Finally, save the results object to disk:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">save</span>(ma,<span class="dt">file=</span><span class="st">&quot;output/ma.rd&quot;</span>, <span class="dt">compress=</span>T)</code></pre>
<p>If “perm.boot” is set to be FALSE, then regular permutations (sampling without replacement) are performed rather than bootstrapping (sampling with replacement).</p>
<p>See the maPerm(), maPermMerge(), and maPermFdr() for manual methods to run permutations. These are useful for spreading permutations across multiple machines or in a high performance computing (HPC) environment.</p>
</div>


</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>