---
layout: default
title: "Goldmine: Integrating information to place epigenome-wide results into biological contexts"
---

      <div class="page-header" id="banner">
        <div class="row">
          <div class="col-lg-8 col-md-7 col-sm-6">
            <h1>Goldmine</h1>
            <p class="lead">Integrating information to place epigenome-wide results into biological contexts</p>
          </div>
          <div class="col-lg-4 col-md-5 col-sm-6">
              <img class="center" src="/assets/img/gm.png" height="200">
          </div>
        </div>
        <div><a href="/goldmine/docs" class="btn btn-default">Read How-To</a>&nbsp;&nbsp;&nbsp;<a href="http://github.com/jeffbhasin/goldmine" class="btn btn-default" target="_blank">View Source on GitHub</a></div>
      </div>

 <div class="container">
    <h1>What is Goldmine?</h1>
    <p>Goldmine is an R package that integrates information to help biologists interpret the genomic ranges that result from epigenome-wide sequencing experiments. The package takes as input any set of genomic ranges (defined by chromosome, start, end) and will annotate these ranges with respect to known genes and features.</p>
    <p><h3>Key Features:</h3>
      <ul>
        <li>Annotates genomic context for sets of genomic ranges (promoter, gene 3' end, exon, intron, intergenic)</li>
        <li>Detailed accounting of specific gene model and feature overlaps (including CpG islands/shores/shelves, COSMIC mutations, GWAS catalog loci, and ENCODE DNaseI sites)</li>
        <li>Downloads and caches the most up-to-date data from the <a href="http://genome.ucsc.edu/">UCSC Genome Browser</a> (including gene sets: UCSC, RefSeq, ENSEMBL; feature sets: ENCODE ChIP-seq supertracks, ENCODE DNaseI supertrack, CpG islands; and all other tables available from UCSC.)</li>
        <li>Works for any species with an assembled and annotated genome available from UCSC</li>
    </ul></p>

    <h1>How to Use Goldmine</h1>
    <p>We have created a step-by-step how to for new users to learn the essential functions of the R package and perform an analysis on example data.</p>
    <p><h3>Step-by-Step How To:</h3>
      <ul>
        <li><a href="/goldmine/docs"><img width="30px" height="30px" src="/assets/img/text-html.svg">View Online</a></li>
        <li><a href="/goldmine/docs/goldmine.pdf"><img width="30px" height="30px" src="/assets/img/application-pdf.svg">Download PDF</a></li>
        <li><a href="/goldmine/docs/goldmine.docx"><img width="30px" height="30px" src="/assets/img/text-rtf.svg">Download DOCX</a></li>
    </ul></p>

    <p>For more advanced options, please refer to the function manual. This information is also available from within R using the "?" command.</p>
    <p><h3>Function Reference Manual:</h3>
      <ul>
        <li><a href="/goldmine/docs/goldmine-manual.pdf"><img width="30px" height="30px" src="/assets/img/application-pdf.svg">Download PDF</a></li>
    </ul></p>

    <h1>Citation</h1>
    <p>Please refer to our open access manuscript to learn more about Goldmine. Please cite this work if Goldmine is useful in your research.</p>
    <p><a href="http://www.ncbi.nlm.nih.gov/pubmed/27257071" target="_blank">Bhasin, JM and Ting, AH. (2016). Goldmine integrates information placing genomic ranges into meaningful biological contexts. Nucleic Acids Res. PMID: 26673711. PMCID: PMC4705678.</a></p>

    <h1>Contact and Help</h1>
    <p>Please open an issue on the <a href="https://github.com/jeffbhasin/goldmine/issues">Goldmine GitHub Issues</a> page for questions, bug reports, or enhancement requests.</p>
    <p>You may also contact the package author with questions or feedback: <script type="text/javascript">eval(unescape("%70%73%75%77%67%34%39%3d%5b%27%25%36%33%25%36%66%25%36%65%25%37%34%25%36%31%25%36%33%25%37%34%27%2c%5b%27%25%36%33%25%36%66%25%36%64%27%2c%27%25%36%61%25%36%35%25%36%36%25%36%36%25%36%32%25%36%39%25%36%66%27%5d%2e%72%65%76%65%72%73%65%28%29%2e%6a%6f%69%6e%28%27%2e%27%29%5d%2e%6a%6f%69%6e%28%27%40%27%29%3b%71%62%75%75%74%31%35%3d%75%6e%65%73%63%61%70%65%28%70%73%75%77%67%34%39%29%3b%64%6f%63%75%6d%65%6e%74%2e%77%72%69%74%65%28%71%62%75%75%74%31%35%2e%6c%69%6e%6b%28%27%6d%61%69%27%2b%27%6c%74%6f%3a%27%2b%70%73%75%77%67%34%39%29%29%3b"));</script></p>

    </div>


